rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+protected by a lock, so only one socket may touch the refcount at a
+time. In practice the fact that there can only be one `RecvChannel`
+per queue means that we can just count `SendChannel` sockets and
+enforce an at-most-one rule for `RecvChannel`. The underlying queue is
+deleted when no socket connections remain. This does mean that if we
+were to e.g. put a message on a queue, then delete all the channels
+before forwarding the message, and then use the UUID to create a new
+`RecvChannel` the queue would be deleted and the message lost. One can
+imagine this scenario happening when navigating a remote context which
+has a `RecvChannel` tied to the UUID in the URL, if the `SendChannel`
+for the context isn't kept alive. Another possibility is to only
+delete the queue if it's empty, and register an explicit completion
+callback in the test that deletes any queue with a UUID known to the
+test context. Of course this doesn't help if a channel is created in a
+non-test context (e.g. for messaging between multiple remotes).
+
+## Example
+
+test.html
+
+```html
+<!doctype html>
+<title>executeScript example</title>
+<script src="/resources/testharness.js">
+<script src="/resources/testharnessreport.js">
+<script src="/resources/channel.js">
+
+<script>
+promise_test(async t => {
+  let remote = new RemoteWindow();
+  window.open(`child.html?uuid=${remote.uuid}`, "_blank", "noopener");
+  let result = await remote.executeScript(id => {
+    return document.getElementById(id).textContent;
+  }, "test");
+  assert_equals("result", PASS);
+});
+</script>
+```
+
+child.html
+
+```html
+<p id="nottest">FAIL</p>
+<p id="test">PASS</p>
+<script>
+start_window();
+</script>
+```
+
+## Possible Future Additions
+
+The primitives here could be integrated more completely with
+testharness.js. For example we could use a `RemoteWindow` as a source
+of tests in `fetch_tests_from_window`. Alternatively, or in addition
+to that, we could integrate asserts with script execution, so that a
+remote context could include a minimal testharness.js and enable
+something like:
+
+```js
+promise_test(t => {
+  let r = new RemoteWindow();
+  window.open(`file.html?uuid=${r.uuid}`, "_blank", "noopener");
+  await r.executeScript(() => {
+    assert_equals(window.opener, null)
+  });
+});
+```
+
+Currently this is possible if `file.html` is put in `resources` or
+`support` directory so it can't be detected as a test. But it would
+perhaps be better to have a minimal subset of testharness available
+without providing the API surface that only makes sense in a test
+window.
+
+It may be possible in the future to replace the backend with a
+WebDriver BiDi based message passing system that would avoid the need
+for websockets. By sticking close to WebDriver BiDi proposed semantics
+the transition may even be seamless.
+
+testdriver integration is possible. For example we could add
+`RemoteContext.testdriver.click` to execute a click in the remote
```

Comment:
Would `RemoteContext.testdriver.click` be a pure JavaScript wrapper like `RemoteContext.executeScript(() => test_driver.click())`?

Reply:
That doesn't work as written, because all testdriver commands have to go via the test window since that's the one that webdriver is using. So it would desugar into `testdriver.click(RemoteContext.uuid)`, and testdriver would learn to look up contexts from a `uuid` parameter in the URL, as well as from the internal identifier it uses today.